### PR TITLE
Clean up `Platform` to be more maintainable

### PR DESCRIPTION
**Pull request by JakeWharton**
Break up functionality into groups to simplify their configuration and make it easier to see how things change on each platform.

For the interface default method handling, replace reflection on newer JVMs with a multi-release jar and version-based variants of the JVM support. In order to sufficiently test this, the Java tests are now in a sibling module similar to Android, Kotlin, and RoboVM so that they consume the final jar rather than exploded class directories.

### CODE REVIEW

File: retrofit/build.gradle [hunk 1]
```diff
@@ -2,6 +2,35 @@ apply plugin: 'java-library'
 apply plugin: 'org.jetbrains.kotlin.jvm'
 apply plugin: 'com.vanniktech.maven.publish'
 
+def addMultiReleaseSourceSet(int version) {
```
Comment (swankjesse) [inline retrofit/build.gradle:5]: This is fantastic power

File: retrofit/java-test/README.md [hunk 2]
```diff
@@ -0,0 +1,6 @@
+# Retrofit Java Tests
+
+These are in a separate module for two reasons:
+
+- It ensures optional dependencies (Kotlin stuff) are completely absent.
+- It uses the multi-release jar on the classpath rather than only the classes folder.
```
Comment (swankjesse) [inline retrofit/java-test/README.md:6]: Thanks for calling this out

File: retrofit/src/main/java/retrofit2/DefaultMethodSupport.java [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+  }
+
+  @Nullable
+  Object invokeDefaultMethod(
+      Method method, Class<?> declaringClass, Object proxy, @Nullable Object[] args)
+      throws Throwable {
+    throw new AssertionError();
+  }
+
+  @TargetApi(24)
```
Comment (swankjesse) [inline retrofit/src/main/java/retrofit2/DefaultMethodSupport.java:41]: Comment why this isn’t `@TargetApi(26)` ?

**Review by swankjesse — APPROVED** [review]

File: retrofit/src/main/java/retrofit2/DefaultMethodSupport.java (hunk 3 above)
Comment (JakeWharton) [reply retrofit/src/main/java/retrofit2/DefaultMethodSupport.java:41]: Done

**Pull request merged**